### DISCUSSION TITLE
Improvements to the current addon/plugin system

### DISCUSSION THREAD

**Discussion by Nukiloco** [godotengine/godot-proposals#4238] [Editor]
# Current issues
Currently Godot's plugin system is already good at its current state but from personal experience there is a few tedious issues that the system has.

## No built-in plugin reloading
Some plugins will error out sometimes or sometimes you need to refresh your plugin code when developing one. Yes there is already an add-on for this, though with Godot supporting plugins outside the box, wouldn't it make sense to add support for this?

## Chaotic addon structures
Some of the files on plugin repositories are above the addon folder which can cause conflict with existing files. For example say that we have this repository
```
addons
  - plugin_name
    - (source code here)
README.md
LICENSE.md
```
When you download this addon from the Godot asset manager, Godot will mark the files README.md and LICENSE.md as conflicting files. Though however I would want to move a file like ``LICENSE.md`` into the plugin_folder itself and I don't think you can move the file before you install the plugin. Some of the repositories sometimes even install files outside the addons folder and can conflict with your existing files.

Other repositories have it setup where all the files would install only in the correct plugin folder.

## Using git urls / other branches, commits, and forks
Sometimes you want to use the latest master version of a plugin repository into your own code. However this can be a little difficult to setup, (aka using git submodules). And with the other issue above, some of the folders installed can go into the wrong folders causing you to have to manually fix up a bunch of relative path issues in scripts.

## Updating addons 
Currently when wanting to install the newest uploaded version of addon/plugin to the asset store, you would have to do the following:
* Go back to the asset store
* Scroll through a lot of pages or manually search for it
* Wait for Godot to update the information on the screen (which does seem to be slow?)
* Select the add-on
* Install the add-on
* Then use the file dialog

In my opinion there should be some sort of plugin manager that allows you to easily to install a new version of a plugin just by a click of one button to update it to any version you want.

## Selecting a version
Currently there is no way to select a particular version of an add on, your forced to use the latest version that was uploaded to the asset store or you would have to manually clone the addon into your repository.

Some people would like to select certain versions of addons because of:
* Add-on has a new code base which would require you to refactor a lot of your game's code
* Current version of the add-on has some bugs that would require a lot of fixing

## Filter add-ons / plugins in the asset store
There is no way to easily to filter addons/plugins when they were recently updated, how many downloads they have, etc...
Currently I would have to manually go to the addons/plugins repository, check how many stars it has to see if the community likes it, check the latest commit in each branch, to see if its actively being updated.

## Inspirations
I got these ideas/inspirations from using javascript ecosystems like [NPM](https://www.npmjs.com/), and [Deno](https://deno.land/). If anyone has any suggestions / feedback / or just a general discussion about the current state of addon management, please leave a comment below! Thanks for reading!

**Comment by aXu-AP** [comment]
+1 about structure. I think asset library should enforce specific folders on anything other than demo projects/templates. #1205 feels like a good solution to this.

Plugin manager definitely feels like needed addition. I'd imagine you have a list of addons downloaded in your computer, and could simply select which ones to add to project (to be clear, the addon would still be copied into project for portability). I feel there's a potential to make a mockup with a plugin. A plugin manager plugin, so to speak :D

**Comment by Calinou** [comment]
See also https://github.com/godotengine/godot-proposals/issues/554.

Note that automatically unchecking files such as top-level `README.md` and `LICENSE.md` was proposed in the past, but it was rejected. It was also proposed to make the file selection dialog not appear if there are no conflicting files, but this was also rejected: https://github.com/godotengine/godot/pull/50132
